log.go

```diff
@@ -182,12 +189,13 @@ func (l *raftLog) findConflictByTerm(index uint64, term uint64) uint64 {
 			index, li)
 		return index
 	}
-	for {
+	for ; ; index-- {
 		logTerm, err := l.term(index)
 		if logTerm <= term || err != nil {
+			// NB: an error can only be ErrCompacted here, but it's always safe to
+			// return the current index, so we don't assert on other errors.
```

Comment:
Yes and no because the method comment claims that this returns the "best possible" index but in the case of a transient storage error this isn't true. Could we return an error here if we see one that is unexpected (i.e. not ErrCompacted).

Reply:
This comment is softer now. Decided not to return an err / overengineer here, because `term()` method currently panics on any error other than `ErrCompacted/ErrUnavailable`. This all feels clunky and not ideal, maybe error handling in this library should be improved.

But also: returning both a higher and a lower hint here is fine - if it doesn't work, there will be another MsgApp iteration, or there will be more entries resent than necessary. The `MsgApp` sender (leader) works in such a way that it can swallow any hint including zero.